master/buildbot/www/oauth2.py

```diff
@@ -206,22 +211,104 @@ def __init__(self,
-            self.resourceEndpoint = '{0}/api/v3'.format(serverURL)
+        self.serverURL = serverURL or self.resourceEndpoint
+
+        if not isinstance(apiVersion, int):
+            config.error(
+                'GitHubAith apiVersion must be of type int not {}'.format(
+                    type(apiVersion)))
+        if apiVersion not in (3, 4):
+            config.error(
+                'GitHubAuth apiVersion must be 3 or 4 not {}'.format(
+                    apiVersion))
+        self.apiVersion = apiVersion
+        if apiVersion == 3:
+            if getTeamsMembership is True:
+                config.error(
+                    'Retrieving team membership information using GitHubAuth is only '
+                    'possible using GitHub api v4.')
+            self.resourceEndpoint = '{0}/api/v3'.format(self.serverURL)
+        else:
+            self.resourceEndpoint = '{0}/graphql'.format(self.serverURL)
+        self.getTeamsMembership = getTeamsMembership
+
+    def post(self, session, query):
+        log.info('POST: {endpoint} -> DATA:\n----\n{data}\n----', endpoint=self.resourceEndpoint, data=query)
+        ret = session.post(self.resourceEndpoint, json={'query': query})
+        return ret.json()
 
     def getUserInfoFromOAuthClient(self, c):
+        if self.apiVersion == 3:
+            return self.getUserInfoFromOAuthClient_v3(c)
+        return self.getUserInfoFromOAuthClient_v4(c)
+
+    def getUserInfoFromOAuthClient_v3(self, c):
         user = self.get(c, '/user')
         emails = self.get(c, '/user/emails')
         for email in emails:
             if email.get('primary', False):
                 user['email'] = email['email']
                 break
         orgs = self.get(c, '/user/orgs')
-
         return dict(full_name=user['name'],
                     email=user['email'],
                     username=user['login'],
                     groups=[org['login'] for org in orgs])
 
+    def getUserInfoFromOAuthClient_v4(self, c):
+        graphql_query = textwrap.dedent('''
+            query {
+              viewer {
+                email
+                login
+                name
+                organizations(first: 100) {
+                  edges {
+                    node {
+                      login
+                    }
+                  }
+                }
+              }
+            }
+        ''')
+        data = self.post(c, graphql_query.strip())
+        data = data['data']
+        log.info('Response: {response}', response=data)
+        user_info = dict(full_name=data['viewer']['name'],
+                         email=data['viewer']['email'],
+                         username=data['viewer']['login'],
+                         groups=[org['node']['login'] for org in
+                                 data['viewer']['organizations']['edges']]
+        )
+        if self.getTeamsMembership:
+            graphql_query = '''
```

Comment:
Invoking jinja just for this? It will be less performant but sure.

Render a string or create a template file and render that?
Do we have helper utilities to prep jinja for rendering?